js_modules/dagster-ui/packages/ui-core/src/assets/RecentUpdatesTimeline.tsx

```diff
@@ -0,0 +1,209 @@
+
+  const bucketedMaterializations = useMemo(() => {
+    if (!viewport.width) {
+      return [];
+    }
+    const firstPassBucketsArray: Array<{
+      number: number;
+      materializations: typeof materializations;
+    }> = new Array(buckets);
+
+    materializations.forEach((materialization) => {
+      const bucketNumber = Math.floor(
+        ((parseInt(materialization.timestamp) - startTimestamp) / range) * buckets,
+      );
+      firstPassBucketsArray[bucketNumber] = firstPassBucketsArray[bucketNumber] || {
+        number: bucketNumber,
+        materializations: [] as typeof materializations,
+      };
+      firstPassBucketsArray[bucketNumber]!.materializations.push(materialization);
+    });
+
+    const secondPassBucketsArray: Array<{
+      start: number;
+      end: number;
+      materializations: typeof materializations;
+    }> = [];
+
+    firstPassBucketsArray.forEach((bucket) => {
+      const lastBucket = secondPassBucketsArray[secondPassBucketsArray.length - 1];
+      if (!lastBucket || lastBucket.end !== bucket.number - 1) {
+        secondPassBucketsArray.push({
+          start: bucket.number,
+          end: bucket.number,
+          materializations: bucket.materializations,
+        });
+      } else {
+        lastBucket.end = bucket.number;
+        lastBucket.materializations = [...lastBucket.materializations, ...bucket.materializations];
+      }
+    });
+    return secondPassBucketsArray;
+  }, [viewport.width, buckets, materializations, startTimestamp, range]);
+
+  const formatDateTime = useFormatDateTime();
+
+  return (
+    <Box flex={{direction: 'column', gap: 4}}>
+      <Box flex={{direction: 'row', justifyContent: 'space-between'}}>
+        <Subtitle2>Recent updates</Subtitle2>
+        <Caption color={Colors.textLighter()}>
+          {materializations.length === 100
+            ? 'Last 100 updates'
+            : `Showing all ${materializations.length} updates`}
+        </Caption>
+      </Box>
+      <Box border="all" padding={6 as any} style={{height: 32}}>
+        <div {...containerProps} style={{width: '100%', height: 20, position: 'relative'}}>
+          {bucketedMaterializations.map((bucket) => {
+            const width = tickWidth * (bucket.end - bucket.start + 1);
+            return (
+              <TickWrapper
+                key={bucket.start}
+                style={{
+                  left: tickWidth * bucket.start,
+                  width,
+                }}
+              >
+                <Popover
+                  content={
+                    <Box flex={{direction: 'column', gap: 8}}>
+                      <Box padding={8} border="bottom">
+                        <Subtitle2>Materializations</Subtitle2>
+                      </Box>
+                      <div style={{maxHeight: 'min(80vh, 300px)', overflow: 'scroll'}}>
+                        {bucket.materializations
+                          .sort((a, b) => parseInt(b.timestamp) - parseInt(a.timestamp))
```

Comment:
Watch out for mutative `sort`

Reply:
ah yeah... I'll sort them in the memo instead